### PR TITLE
GPX export fixes for heartrate

### PR DESCRIPTION
**Pull request by andrewheadricke**
This PR incorporates a minimal set of tweaks to make heartrate export work on amazfit BIP, particularly for upload to Strava.com.

Changes include:
* Merge some of @swiss5555 changes relating to heart rate tracking
* Change the GPX XML to not output the src on every track point.
* Make sure the heart rate value is wrapped in a `TrackPointExtension` element.

I fully expect this PR to require some review before merging :)

### CODE REVIEW

**Comment by danielegobbetti** [comment]
Do you know any open source GPX visualizer/editor that shows HR data? I am unable to test your code without one.

File: app/src/main/java/nodomain/freeyourgadget/gadgetbridge/export/GPXExporter.java [hunk 1]
```diff
@@ -88,7 +90,7 @@ private void exportTrack(XmlSerializer ser, ActivityTrack track) throws IOExcept
         String source = getSource(track);
         boolean atLeastOnePointExported = false;
         for (ActivityPoint point : trackPoints) {
-            atLeastOnePointExported = exportTrackPoint(ser, point, source) | atLeastOnePointExported ;
+            atLeastOnePointExported = exportTrackPoint(ser, point, source, trackPoints) | atLeastOnePointExported ;
```
Comment (cpfeiffer) [inline app/src/main/java/nodomain/freeyourgadget/gadgetbridge/export/GPXExporter.java]: Good catch! I'll change the assignment to `|=` which is a bit more common.

File: app/src/main/java/nodomain/freeyourgadget/gadgetbridge/export/GPXExporter.java [hunk 2]
```diff
@@ -117,29 +119,53 @@ private boolean exportTrackPoint(XmlSerializer ser, ActivityPoint point, String
         if (description != null) {
             ser.startTag(NS_DEFAULT, "desc").text(description).endTag(NS_DEFAULT, "desc");
         }
-        ser.startTag(NS_DEFAULT, "src").text(source).endTag(NS_DEFAULT, "src");
+        //ser.startTag(NS_DEFAULT, "src").text(source).endTag(NS_DEFAULT, "src");
```
Comment (cpfeiffer) [inline app/src/main/java/nodomain/freeyourgadget/gadgetbridge/export/GPXExporter.java]: Why did you remove the `src` element?

File: app/src/main/java/nodomain/freeyourgadget/gadgetbridge/export/GPXExporter.java [hunk 3]
```diff
@@ -117,29 +119,53 @@ private boolean exportTrackPoint(XmlSerializer ser, ActivityPoint point, String
...
+    private void exportTrackpointExtensions(XmlSerializer ser, ActivityPoint point, List<ActivityPoint> trackPoints) throws IOException {
         if (!includeHeartRate) {
             return;
         }
 
         int hr = point.getHeartRate();
         if (!HeartRateUtils.isValidHeartRateValue(hr)) {
+            if(!includeHeartRateByTime) { return; }
+
+            Date time = point.getTime();
```
Comment (cpfeiffer) [inline app/src/main/java/nodomain/freeyourgadget/gadgetbridge/export/GPXExporter.java]: I'll try to factor this out into a separate method. Unfortunately it's still O(n^2), but we can fix this later.

File: app/src/main/java/nodomain/freeyourgadget/gadgetbridge/export/GPXExporter.java [hunk 4]
```diff
@@ -117,29 +119,53 @@ private boolean exportTrackPoint(XmlSerializer ser, ActivityPoint point, String
...
+                    }
+                }
+            }
+            if(closestPointItem != null) {
+                hr = closestPointItem.getHeartRate();
+                if (!HeartRateUtils.isValidHeartRateValue(hr)) {
+                    return;
+                }
+            }
             return;
```
Comment (cpfeiffer) [inline app/src/main/java/nodomain/freeyourgadget/gadgetbridge/export/GPXExporter.java]: BTW, this return here made the whole nearest sample finding useless, AFAICS.

File: app/src/main/java/nodomain/freeyourgadget/gadgetbridge/export/GPXExporter.java (hunk 2 above)
Comment (andrewheadricke) [reply app/src/main/java/nodomain/freeyourgadget/gadgetbridge/export/GPXExporter.java]: Because a GPX file with 1000 track points would output "Amazfit BIP......." 1000 times.. It would make more sense to put the src in the header just once.  But just for expediency I simply commented it out.

File: app/src/main/java/nodomain/freeyourgadget/gadgetbridge/export/GPXExporter.java (hunk 4 above)
Comment (andrewheadricke) [reply app/src/main/java/nodomain/freeyourgadget/gadgetbridge/export/GPXExporter.java]: Quite possible, I did notice that the HR samples were still only appearing every 5th track point.  I didn't get a chance to dig any further.

**Comment by andrewheadricke** [comment]
@danielegobbetti I just tried https://github.com/tumic0/GPXSee and it works pretty well.. shows heart rate just fine.
